### ISSUE TITLE
Decentralized dropbox over Freenet

### ISSUE THREAD

**Issue by sanity** [freenet/freenet-core#680]
[Unison](https://github.com/bcpierce00/unison) is an open source dropbox replacement that works over ssh to synchronize files and directories. It should be straightforward to build something similar on locutus that's even easier to use, something like:

```bash
$ freesync share code/freenet/locutus
code/freenet/locutus now shared, secret token: MFZWIZTBONSGMYLTMRTGC43EOZQXGZDWMFZWI5TW
```

On a different machine:

```bash
$ freesync add MFZWIZTBONSGMYLTMRTGC43EOZQXGZDWMFZWI5TW Documents/code/locutus
Directory synced to Documents/code/locutus
```

And you've got a cloudless dropbox just like that.

**Comment by iduartgomez** [comment]
Duplicated?

https://github.com/freenet/locutus/issues/407

**Comment by sanity** [comment]
🤖 **Relabeled for consistency**

Applied labels:
- `T-feature`
- `P-medium`
- `E-medium`
- `A-networking`
- `S-needs-design`

**Reasoning:** The issue proposes a new decentralized file-sync feature (T-feature). It is not urgent or release-blocking, so P-medium is appropriate. Implementing reliable peer-to-peer file synchronization involves moderate complexity (E-medium) and primarily concerns networking and user-facing tooling (A-networking, A-developer-xp). The proposal is high-level and will require design work before implementation (S-needs-design). Confidence values reflect certainty about type and area and slightly less certainty about implementation effort and priority.

Previous labels: `help wanted,T-feature`

**Comment by sanity** [comment]
# Issue Priority Assessment

Generated: 2026-06-09T14:10:24.638925093+00:00
Issue hash: `dab2010f`

## Value Assessment

This issue is a feature proposal for a decentralized file-synchronization tool ("freesync") built on top of Freenet. Here is the assessment:

**Relevance Check:** This issue remains relevant in the sense that it has not been implemented — there is no `freesync` tool, no file-sync application, and no supporting infrastructure in the current codebase. The `apps/` directory contains only `freenet-ping` as an example application. However, the issue is flagged as a potential duplicate of locutus#407, suggesting this idea has been floated before without progressing. The feature is entirely unbuilt, so it is neither obsolete nor already resolved — it is simply unstarted.

**User Impact:** The impact would be meaningful for a subset of privacy-conscious users who want Dropbox-like sync without trusting a centralized cloud provider. However, this audience is speculative; Freenet currently has no user base for this kind of application, and the primary users today are developers and early testers. The feature would only become impactful after Freenet itself achieves stable operation at scale — so the practical user impact right now is near zero, contingent on many preconditions being met first.

**Strategic Alignment:** The feature aligns with Freenet's long-term vision of decentralized applications but is far removed from the current development focus. The codebase is still maturing the core protocol (transport, DHT routing, operation state machines, contract/delegate WASM execution). Building a reliable file-sync application on top requires those primitives to be stable, performant, and well-tested at network scale — none of which is fully true yet. The `S-needs-design` label is apt: the issue gives no design for how large files would be sharded, how conflict resolution would work, how the secret token maps to access control, or how deletion/rename semantics would behave in a decentralized setting. These are non-trivial problems.

**Blocking Factor and Urgency:** This blocks nothing in the current development pipeline. It is a greenfield application idea with no dependencies running in the other direction. There is no time-sensitive reason to address it soon — it should wait until the core protocol is stable enough to support application development reliably. The medium priority assigned by the label bot seems slightly generous; given the design gaps, the duplicate flag, and the distance from current engineering priorities, this is realistically a long-horizon roadmap item that will require substantial design work before any implementation begins.

## Cost Assessment

This issue proposes building a "freesync" tool — a decentralized, cloudless Dropbox/Unison alternative on top of Freenet. No such implementation exists in the repository; it would be a greenfield application built on the contract/delegate infrastructure.

The scope is large but fully in userspace: it would live as a new app under `apps/freesync/` (mirroring the structure of `apps/freenet-ping/`), consisting of a WASM contract for distributed state management and a CLI tool for the `freesync share` / `freesync add` / file-watching loop. No changes to core freenet-core internals are required. However, the problem is substantially harder than the issue implies. A file sync system needs to handle: efficient binary chunking (files can be gigabytes, but Freenet contracts store state as blobs), deduplication of content blocks, CRDT-compatible conflict resolution across concurrent edits on multiple machines, secret/key derivation for the share token, and incremental delta transfer using the `get_state_delta` / `summarize_state` contract interface. Each of these is non-trivial. The contract state model (commutative, associative merges, per-key size limits) is well-suited to small metadata structures but would need careful design for storing large file trees — likely a two-layer design with a metadata contract plus content-addressed block contracts.

There are no external dependencies in the bureaucratic sense (no third-party services, keys to acquire, or organizational approvals), but the design work is substantial. The issue itself is labeled `S-needs-design` for good reason: the naive approach of storing file contents directly in contract state would hit size and bandwidth limits immediately. A proper design requires a content-addressed storage layer (chunking files into fixed-size blocks, each stored as a separate contract keyed by content hash), a metadata manifest contract, and a sync protocol that avoids retransmitting blocks the remote already has. The `summarize_state` + `get_state_delta` primitives exist precisely for this but have never been exercised at the scale of a real file sync workload. The mention in a comment that this may duplicate issue #407 further suggests this has been in design limbo for years.

Risk to existing behavior is low — this is purely additive and isolated from core routing, transport, and operation state machines. Testing, however, is expensive: a file sync tool needs correctness tests for conflict resolution, property tests for CRDT merge commutativity, integration tests exercising multi-node sync across the simulation framework, and performance benchmarks for chunking large files. The affected code is easy to unit test in isolation, but end-to-end testing across multiple simulated nodes (using `run_simulation_direct`) would be required to validate that sync actually converges under realistic network conditions. Overall, this is a high-effort feature — likely 4–8 weeks of focused engineering — with the design phase being the dominant cost before a single line of implementation is written.

---
*Generated by issue-prioritizer*

**Comment by sanity** [comment]
🤖 **Automated ranking by Torvan.** Not a maintainer's assessment of this issue.

| Measure | Score |
|---|---:|
| Value | 0.039 |
| Cost | 0.932 |
| Priority | 588 of 593 |

Value and cost run 0 to 1 across the 593 open issues in the freenet repositories: value 1.00 is the most valuable, cost 0.00 the cheapest. Issues are compared against each other rather than scored in isolation. The queue alternates between the most valuable work and the best value for the effort, so neither expensive-but-important nor cheap-and-quick work crowds the other out.

Ranked 2026-08-31, and liable to change. Adding information here, whether that is scope, a reproduction, or a correction to this assessment, causes the issue to be reassessed and this comment updated.

**Comment by glesage** [comment]
### Project-board closure review

This appears to duplicate the older open issue [#407 — Directory synchronization service / decentralized Dropbox](https://github.com/freenet/freenet-core/issues/407).

Both issues propose the same product: securely synchronizing files and directories across machines over Freenet as a cloudless Dropbox replacement. This issue adds a possible freesync CLI sketch, but not a distinct requirement or incompatible design.

I think this can be closed as a duplicate and the CLI example preserved on #407 so discussion has one canonical home. I have not closed it; this comment is for maintainer confirmation.
